### ISSUE TITLE
DIFFERENT ADDRESS - problem with duplicit ident field

### ISSUE THREAD

**Issue by milo** [iputils/iputils#489]
Hi. I'm aware that this is probably not an issue of this package, but I don't really know who to ask. So if you find it inappropriate please close the issue, I'll understand.

I'm running Debian 12 (Bookworm) with bundled kernel 6.1.0-11-amd64. I'm running Nagios 4.4.6-4 and as a `check-host-alive` I use your `ping` in version 20221126-1. I tried bundled kernel 5.10.0-25-amd64 with the same result.

Before upgrading to this setup, I used Debian 11 with kernel v4. There where no issues.

As a `check-host-alive` I use command:
```
/bin/ping4 -n -v -D -W 1 -i 1 -c 5 -M 'do' -s 56 -O "$Host" 2>&1
```
and when everything OK, I'll get (IP obfuscated):
```
/bin/ping4: sock4.fd: 3 (socktype: SOCK_RAW), sock6.fd: -1 (socktype: 0), hints.ai_family: AF_INET

ai-ai_family: AF_INET, ai-ai_canonname: 10.66.a.b
PING 10.66.a.b (10.66.a.b) 56(84) bytes of data.
[1695120545.794919] 64 bytes from 10.66.a.b: icmp_seq=1 ident=59225 ttl=63 time=0.563 ms
[1695120546.796904] 64 bytes from 10.66.a.b: icmp_seq=2 ident=59225 ttl=63 time=0.557 ms
[1695120547.826579] 64 bytes from 10.66.a.b: icmp_seq=3 ident=59225 ttl=63 time=0.532 ms
[1695120548.854606] 64 bytes from 10.66.a.b: icmp_seq=4 ident=59225 ttl=63 time=0.543 ms
[1695120549.874664] 64 bytes from 10.66.a.b: icmp_seq=5 ident=59225 ttl=63 time=0.540 ms

--- 10.66.a.b ping statistics ---
5 packets transmitted, 5 received, 0% packet loss, time 4080ms
rtt min/avg/max/mdev = 0.532/0.547/0.563/0.011 ms

Date/Time: Tue Sep 19 12:49:09 CEST 2023

```

But time to time, and that's the issue, I get `DIFFERENT ADDRESS` responses:
```
/bin/ping4: sock4.fd: 3 (socktype: SOCK_RAW), sock6.fd: -1 (socktype: 0), hints.ai_family: AF_INET

ai-ai_family: AF_INET, ai-ai_canonname: 10.66.a.b
PING 10.66.a.b (10.66.a.b) 56(84) bytes of data.
[1695120483.778898] 64 bytes from 10.66.a.b: icmp_seq=1 ident=12506 ttl=63 time=0.528 ms
[1695120483.794459] 64 bytes from 147.32.c.d: icmp_seq=3 ident=12506 ttl=63 time=0.439 ms (DIFFERENT ADDRESS!)
[1695120484.780130] 64 bytes from 10.66.a.b: icmp_seq=2 ident=12506 ttl=63 time=0.542 ms
[1695120484.793663] 64 bytes from 147.32.c.d: icmp_seq=4 ident=12506 ttl=63 time=0.384 ms (DIFFERENT ADDRESS!)
[1695120485.781270] 64 bytes from 10.66.a.b: icmp_seq=3 ident=12506 ttl=63 time=0.531 ms

--- 10.66.a.b ping statistics ---
3 packets transmitted, 5 received, -66.6667% packet loss, time 2002ms
rtt min/avg/max/mdev = 0.384/0.484/0.542/0.062 ms

Date/Time: Tue Sep 19 12:48:05 CEST 2023
```
As you can see, I'll get response from different IP but with **identical** `ident` field. The different IP adress is monitored by Nagios too and I can sniff echo requests invokend by Nagios for this IP in the same time. I monitor 582 hosts in meantime.

Btw. note the negative packet loss :)

I don't know how echo reply is paired with echo request, but seems to me, that via `ident` field. I also don't know what is responsible to generate `ident` field. I found a little bit about topic - the application can set it... or "set it zero, kernel will set it, or... use PID... - probably a little bit true from all the statements.

I found the kernel 6 and ident comment but I'm not so C advanced programmer to track it somehow

https://github.com/iputils/iputils/blob/8d1acb6d95731ce1b2c7ee1c361bccfaa0918487/ping/ping.c#L587-L590

I found `-e` parameter description in man pages and I'll probably use it. But I'm really curious if there is another way, some OS-wide way like some `/sys/class/net/...` setting or similar to change `ident` generator behaviour. More "randomness" or predictable "sequenceness"...

Many thanks for your time you spend on this.

**Comment by pevik** [comment]
Debian 11 Bullseye (oldstable) uses older iputils 20210202, could you test it on your upgraded Debian 12 Bookworm (stable)? That would show if the change comes from kernel (more likely) or iputils. If the problems is in iputils and you are able to compile and use it temporarily, you'd probably be able to bisect it to find the commit which caused the change. 

Also, could you please describe the network setup you use? It would be great if I was able to setup some testing environment with VM, because negative packet loss is probably a bug (unless somebody detect DIFFERENT ADDRESS with this -).

There were various changes in later kernels, which improved ICMP datagram socket (`SOCK_DGRAM`). But Debian still uses raw socket (`SOCK_RAW`), due non-systemd init systems still being supported (see https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1008281#71 and https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1027773), so it should not affect it. If you want to try ICMP datagram socket (it might work differently), you can enable it with:

```
$ sysctl net.ipv4.ping_group_range # prints your current value
net.ipv4.ping_group_range = ... 

# sysctl net.ipv4.ping_group_range="1 1000"
```

To go back to raw socket you run:
```
sysctl net.ipv4.ping_group_range="0 1"
```
Obviously setup done like this is just until reboot (it needs to go to some file in `/etc/sysctl.d/*.conf`).

**Comment by milo** [comment]
Thank you for your reply! I'll try older iputils and I'll try the ping_group_range setting. In the meantime, here's my network setup.

We use about 140 VLANs with private or public IPv4/IPv6 networks. Hosts are monitored by Nagios across these networks and few IP adresses out of our networks. But Nagios server has only net interfaces in few "special" VLANs and the most of traffic is routed. The content of `/etc/network/interfaces` follows. Btw. I was suspisious about the "old fashion" vs. new systemd network setup so I tried network setup via `systemd-networkd` but it has no impact on the ping issue.

```
auto lo
iface lo inet loopback

#
# eno1 - 1Gb
#
auto eno1


# main interface - most of traffic
auto eno1.900
iface eno1.900 inet static
        address 147.32.x.x
        netmask 255.255.255.128
        gateway 147.32.x.1


iface eno1.900 inet6 static
    address 2001:718:2:xxxx::x:x
    netmask 64
    gateway 2001:718:2:xxxx::1


# management, not routed at all
auto eno1.699
iface eno1.699 inet static
        address 10.0.x.x
        netmask 255.192.0.0
        pre-up echo 1 > /proc/sys/net/ipv6/conf/$IFACE/disable_ipv6



#
# ens1f1 - 10Gb
#
auto ens1f1
iface ens1f1 inet manual
        post-up ip link set $IFACE mtu 9000


# Storage - NFS - not routed at all
auto ens1f1.972
iface ens1f1.972 inet static
        address 172.20.x.x
        netmask 255.255.255.0
        pre-up echo 1 > /proc/sys/net/ipv6/conf/$IFACE/disable_ipv6
        post-up ip link set $IFACE mtu 9000


# Storage - CIFS - not routed at all
auto ens1f1.973
iface ens1f1.973 inet static
        address 172.20.y.y
        netmask 255.255.255.0
        pre-up echo 1 > /proc/sys/net/ipv6/conf/$IFACE/disable_ipv6
        post-up ip link set $IFACE mtu 9000


# not routed at all
auto ens1f1.974
iface ens1f1.974 inet static
        address 172.21.x.x
        netmask 255.255.255.0
        pre-up echo 1 > /proc/sys/net/ipv6/conf/$IFACE/disable_ipv6
        post-up ip link set $IFACE mtu 9000
```

There are no other setups in `/etc/sysctl.d` so pretty default installation.

**Comment by milo** [comment]
I tried 20210202 version and it works fine. I think that's because of 5f6bec5ab57cc8beaa78f5756a0ffbdf01f28d36. There can be ident duplicates but ping 20210202 just ignores them. I agree that showing them is better but should it affects statistics like packet loss? For example, I think that packet loss should be 100% if only "DIFFERENT ADDRESS" packets are received. Probably same with "DUP". But it is only my personal point of view.

So I've played with newer ping 20221126 and `ping_group_range` to enforce `SOCK_DGRAM` usage. With `SOCK_DGRAM` ping utility behaves differently. In one terminal I've run `ping -v 10.0.0.1` and ensures it is `SOCK_DGRAM` socket type. In a second terminal I've sniffed for ident (because ident is not printed in `-v` mode) and then run `ping -v -e $ident 10.0.0.1`. The ping in 1st terminal did not print `DIFFERENT ADDRESS`, only stopped to print anything just like with packet loss. I have no conclusion, just want to mention that.

But back to the original issue - a duplicate ident field. I monitor 582 hosts every minute by Nagios and seems to me that `rand()` is somehow poor if collision can occure. Even if all 582 checks run in the same time, average is about 75 checks at onece, a collision seems unlikely to me.

**Comment by pevik** [comment]
> I've sniffed for ident (because ident is not printed in `-v` mode)

Thanks for your bug report. This is unrelated, but I posted a fix: https://github.com/iputils/iputils/pull/490
Would you mind to share your email address, so that I can add it to `Reported-by:` tag (there is form `Reported-by:` Your Name <your@email>`, see https://www.kernel.org/doc/html/latest/process/submitting-patches.html#sign-your-work-the-developer-s-certificate-of-origin).

**Comment by pevik** [comment]
> But back to the original issue - a duplicate ident field. I monitor 582 hosts every minute by Nagios and seems to me that `rand()` is somehow poor if collision can occure. Even if all 582 checks run in the same time, average is about 75 checks at onece, a collision seems unlikely to me.

First, do these 582 checks run from the same machine? I suppose so, i.e. Nagios runs 582x `ping` every minute right?

Here is random numbers init code:
https://github.com/iputils/iputils/blob/8d1acb6d95731ce1b2c7ee1c361bccfaa0918487/iputils_common.c#L113-L152

IMHO all linux distros should have `HAVE_GETRANDOM` defined, thus it uses [`getrandom(2)`](https://man7.org/linux/man-pages/man2/getrandom.2.html), which is used for seed (set by [`srand(3)`](https://linux.die.net/man/3/srand) + then a different number of `rand()` calls is run (up to 31) => this should be sufficient to get a good enough random number (IMHO more random than using what previously `ping` used - `$PID` - changed in 5026c2221a15bf13e601eade015c971bf07a27e9). But I'll ask others recheck that code, maybe I miss something.

I also wonder if you put debugging like this: 

```
+++ iputils_common.c
@@ -114,6 +114,7 @@ static unsigned int iputil_srand_fallback(void)
 {
        struct timespec ts;
 
+       printf("run iputil_srand_fallback()\n");
        clock_gettime(CLOCK_REALTIME, &ts);
        return ((getpid() << 16) ^ getuid() ^ ts.tv_sec ^ ts.tv_nsec);
 }
@@ -128,6 +129,7 @@ void iputils_srand(void)
        do {
                errno = 0;
                ret = getrandom(&i, sizeof(i), GRND_NONBLOCK);
+               printf("errno: %d\n", errno);
                switch (errno) {
                case 0:
                        break;
@@ -143,6 +145,7 @@ void iputils_srand(void)
        i = iputil_srand_fallback();
 #endif
        srand(i);
+       printf("srand(%d)\n", i);
        /* Consume up to 31 random numbers */
        i = rand() & 0x1F;
        while (0 < i) {
```
You will always get `errno: 0` (i.e. there is no`11` (`EAGAIN`) which would mean not enough entropy - e.g. shortly after boot or if another process eats it) and different value in `srand()`. With `fprintf(stderr, ...)` it could be even logged to stderr, which might occur in some log, so that we could really check if there was a collision.

**Comment by pevik** [comment]
BTW `check-host-alive` command comes from [`monitoring-plugins-basic`](https://packages.debian.org/bookworm/monitoring-plugins-basic) package (because [Debian specific downstream patch](https://sources.debian.org/patches/nagios4/4.4.6-4/70_remove_check-host-alive.patch/) overwrites `nagios4` package), which requires `ping` from iputils (us): [`dep: iputils-ping`](https://packages.debian.org/bookworm/iputils-ping). I wonder whether `ping` from [`inetutils-ping`](https://packages.debian.org/search?keywords=inetutils-ping) or from [`busybox`](https://packages.debian.org/sid/busybox) would perform better.

Busybox uses only raw socket and also PID.

```C
G.myid = (uint16_t) getpid();
/* we can use native-endian ident, but other Unix ping/traceroute
 * utils use *big-endian pid*, and e.g. traceroute on our machine may be
 * *not* from busybox, idents may collide. Follow the convention:
 */
G.myid = htons(G.myid);

```
https://git.busybox.net/busybox/tree/networking/ping.c#n376

I'll have look also on ping from inetutils.

BTW I realized that ICMP datagram socket could also use random ident (just like raw socket). Could you please have look at https://github.com/iputils/iputils/pull/490?

**Comment by milo** [comment]
> Would you mind to share your email address

Sure, if you find it convenient. It is miloslav.hula@gmail.com

**Comment by milo** [comment]
> BTW `check-host-alive` command comes from [`monitoring-plugins-basic`](https://packages.debian.org/bookworm/monitoring-plugins-basic) package

That's where it all started. I used to use `check-host-alive` as defined in distro's monitoring plugins, it is default Nagios host alive check, but after upgrade to Bookwork, it started to behave buggy so I started to investigate. Default `check-host-alive` definition uses `check_ping` binary which is only [wrapper for](https://github.com/monitoring-plugins/monitoring-plugins/blob/master/plugins/check_ping.c) `ping` command.

After upgrade to Bookworm, some hosts became into UP state only for minute in Nagios even we were absolutely sure it is not possible. So I've changed default `check-host-command` definition to my own, as described in this issue description, to get more verbose output:
```
#!/bin/bash
...
/bin/ping4 -n -v -D -W 1 -i 1 -c 5 -M 'do' -s 56 -O "$Host" 2>&1
...
... some stdout parsing
```

After that, I discovered `DIFFERENT ADDRESS` reply with duplicate ident field. I did not investigate too deep but I guess, Nagios native `check_ping` wrapper wrongly parse `ping` output and resolves to `OK` state.

When thinking about it, this is probably Nagios `check_ping` issue too but I'm really curious what is the poor random ident background.

**Comment by milo** [comment]
> First, do these 582 checks run from the same machine? I suppose so, i.e. Nagios runs 582x `ping` every minute right?

Yes, correct. The same machine, good old hardware, not virtual one. 582 ping every minute but in average, 75-100 pings in parallel.

Part of the `ps axuf` tree:
```
nagios   2031563  0.0  0.0  17028  6680 ?        S    Sep21   0:13  \_ /usr/sbin/nagios4 --worker /var/lib/nagios4/rw/nagios.qh
nagios    310404  0.0  0.0   9464  3408 ?        S    08:54   0:00  |   \_ /bin/bash /usr/local/bin/nagios/check_host_alive -H 10.13.a.b
nagios    310405  0.0  0.0   9464  1824 ?        S    08:54   0:00  |   |   \_ /bin/bash /usr/local/bin/nagios/check_host_alive -H 10.13.a.b
nagios    310406  0.0  0.0  10068  1096 ?        S    08:54   0:00  |   |       \_ /bin/ping4 -n -v -D -W 1 -i 1 -c 5 -M do -s 56 -O 10.13.a.b
nagios    310911  0.0  0.0   9464  3428 ?        S    08:54   0:00  |   \_ /bin/bash /usr/local/bin/nagios/check_host_alive -H 147.32.c.d
nagios    310919  0.0  0.0   9464  1716 ?        S    08:54   0:00  |       \_ /bin/bash /usr/local/bin/nagios/check_host_alive -H 147.32.c.d
nagios    310932  0.0  0.0  10068  1104 ?        S    08:54   0:00  |           \_ /bin/ping4 -n -v -D -W 1 -i 1 -c 5 -M do -s 56 -O 147.32.c.d
nagios   2031564  0.0  0.0  17028  6672 ?        S    Sep21   0:13  \_ /usr/sbin/nagios4 --worker /var/lib/nagios4/rw/nagios.qh
nagios    310903  0.0  0.0   9464  3368 ?        S    08:54   0:00  |   \_ /bin/bash /usr/local/bin/nagios/check_host_alive -H 147.32.e.f
nagios    310904  0.0  0.0   9464  1708 ?        S    08:54   0:00  |       \_ /bin/bash /usr/local/bin/nagios/check_host_alive -H 147.32.e.f
nagios    310905  0.0  0.0  10068  1120 ?        S    08:54   0:00  |           \_ /bin/ping4 -n -v -D -W 1 -i 1 -c 5 -M do -s 56 -O 147.32.e.f
nagios   2031565  0.0  0.0  17028  6608 ?        S    Sep21   0:13  \_ /usr/sbin/nagios4 --worker /var/lib/nagios4/rw/nagios.qh
nagios    310882  0.0  0.0   9464  3400 ?        S    08:54   0:00  |   \_ /bin/bash /usr/local/bin/nagios/check_host_alive -H 147.32.g.h
nagios    310894  0.0  0.0   9464  1860 ?        S    08:54   0:00  |       \_ /bin/bash /usr/local/bin/nagios/check_host_alive -H 147.32.g.h
nagios    310896  0.0  0.0  10068  1108 ?        S    08:54   0:00  |           \_ /bin/ping4 -n -v -D -W 1 -i 1 -c 5 -M do -s 56 -O 147.32.g.h
nagios   2031566  0.0  0.0  17028  6628 ?        S    Sep21   0:13  \_ /usr/sbin/nagios4 --worker /var/lib/nagios4/rw/nagios.qh
nagios    310425  0.0  0.0   9464  3340 ?        S    08:54   0:00  |   \_ /bin/bash /usr/local/bin/nagios/check_host_alive -H 147.32.i.j
nagios    310426  0.0  0.0   9464  1776 ?        S    08:54   0:00  |   |   \_ /bin/bash /usr/local/bin/nagios/check_host_alive -H 147.32.i.j
nagios    310427  0.0  0.0  10068  1132 ?        S    08:54   0:00  |   |       \_ /bin/ping4 -n -v -D -W 1 -i 1 -c 5 -M do -s 56 -O 147.32.i.j
nagios    310910  0.0  0.0   9464  3360 ?        S    08:54   0:00  |   \_ /bin/bash /usr/local/bin/nagios/check_host_alive -H 147.32.k.l
nagios    310918  0.0  0.0   9464  1712 ?        S    08:54   0:00  |       \_ /bin/bash /usr/local/bin/nagios/check_host_alive -H 147.32.k.l
nagios    310927  0.0  0.0  10068  1092 ?        S    08:54   0:00  |           \_ /bin/ping4 -n -v -D -W 1 -i 1 -c 5 -M do -s 56 -O 147.32.k.l
nagios   2031567  0.0  0.0  17028  6624 ?        S    Sep21   0:13  \_ /usr/sbin/nagios4 --worker /var/lib/nagios4/rw/nagios.qh
nagios    310440  0.0  0.0   9464  3408 ?        S    08:54   0:00  |   \_ /bin/bash /usr/local/bin/nagios/check_host_alive -H 10.10.m.n
nagios    310441  0.0  0.0   9464  1680 ?        S    08:54   0:00  |   |   \_ /bin/bash /usr/local/bin/nagios/check_host_alive -H 10.10.m.n
nagios    310442  0.0  0.0  10068  1108 ?        S    08:54   0:00  |   |       \_ /bin/ping4 -n -v -D -W 1 -i 1 -c 5 -M do -s 56 -O 10.10.m.n
nagios    310906  0.0  0.0   9464  3428 ?        S    08:54   0:00  |   \_ /bin/bash /usr/local/bin/nagios/check_host_alive -H 147.32.o.p
nagios    310921  0.0  0.0   9464  1704 ?        S    08:54   0:00  |       \_ /bin/bash /usr/local/bin/nagios/check_host_alive -H 147.32.o.p
nagios    310926  0.0  0.0  10068  1128 ?        S    08:54   0:00  |           \_ /bin/ping4 -n -v -D -W 1 -i 1 -c 5 -M do -s 56 -O 147.32.o.p
```

**Comment by milo** [comment]
> You will always get `errno: 0` (i.e. there is no`11` (`EAGAIN`) which would mean not enough entropy - e.g. shortly after boot or if another process eats it) and different value in `srand()`. With `fprintf(stderr, ...)` it could be even logged to stderr, which might occur in some log, so that we could really check if there was a collision.

I tried with stderr output and run it in loop:

```
`for i in $(seq 1 1000); do builddir/ping/ping ... >>/tmp/ping.log; done
```
which I hope emulates Nagios processes. And I found 8 collisions on first run, 9 on second run.

The `rand()` itself is not a collision source. Even with 10000 loop run. But the truncating is:
```c
    if (sock->socktype == SOCK_RAW && rts->ident == -1) {
        int tmp;
        tmp = rand();
        rts->ident = tmp & IDENTIFIER_MAX;  // <<<< source of collisions
        fprintf(stderr, "rand = %u\n", tmp);
        fprintf(stderr, "rts->ident = %d\n", rts->ident);
    }
```

Thinking out loud... there could always be collisions. Much sooner before Bookworm upgrades. But because `ping` used to mute `DIFFERENT ADDRESS` states, Nagios native `check_ping` binary never noticed.

The `rts->ident` is 2 bytes wide, so to ensure 65535 random numbers with the eqal probability distribution is a hard task I think. Maybe for such purpose like this - thousands ping per minute - is better to use cyclic sequence in userspace. Or bring other mechanism how to pair echo request with reply. Some random payload maybe... anyway - when I switch to `SOCK_DGRAM` and kernel assigns identifiers, everything works for me.

**Comment by pevik** [comment]
> The rand() itself is not a collision source. Even with 10000 loop run. But the truncating is:
...

> Thinking out loud... there could always be collisions. Much sooner before Bookworm upgrades. But because `ping` used to mute `DIFFERENT ADDRESS` states, Nagios native `check_ping` binary never noticed.
> 
> The `rts->ident` is 2 bytes wide, so to ensure 65535 random numbers with the eqal probability distribution is a hard task I think. Maybe for such purpose like this - thousands ping per minute - is better to use cyclic sequence in userspace. Or bring other mechanism how to pair echo request with reply. Some random payload maybe... anyway - when I switch to `SOCK_DGRAM` and kernel assigns identifiers, everything works for me.

Ah, you right, I overlook that. Obviously, random number is much higher (`rand()` retuns int), but yes, we need to truncate by `0xFFFF`, because we are limited by range <0, 65535>.

BTW of course, all implementations do it:
[busybox](https://git.busybox.net/busybox/tree/networking/ping.c#n376) by forcing size `uint16_t` and casting:

```C
uint16_t myid;
...
myid = (uint16_t) getpid();
```

and inetutils the same way as us via bitwise & ([IPv4](https://git.savannah.gnu.org/cgit/inetutils.git/tree/ping/libping.c#n119), [IPv6](https://git.savannah.gnu.org/cgit/inetutils.git/tree/ping/ping6.c#n890)):

```C
/* Make sure we use only 16 bits in this field, id for icmp is a unsigned short.  */
p->ping_ident = ident & 0xFFFF;
```

Busybox even [mentions](https://git.busybox.net/busybox/tree/networking/ping.c#n377) possible collisions, but due endianity:

```C
/* we can use native-endian ident, but other Unix ping/traceroute
 * utils use *big-endian pid*, and e.g. traceroute on our machine may be
 * *not* from busybox, idents may collide. Follow the convention:
 */
myid = htons(myid);
```

**Comment by pevik** [comment]
> Thinking out loud... there could always be collisions. Much sooner before Bookworm upgrades. But because `ping` used to mute `DIFFERENT ADDRESS` states, Nagios native `check_ping` binary never noticed.

And your older comment
> I tried 20210202 version and it works fine. I think that's because of 5f6bec5ab57cc8beaa78f5756a0ffbdf01f28d36. There can be ident duplicates but ping 20210202 just ignores them. I agree that showing them is better but should it affect statistics like packet loss? For example, I think that packet loss should be 100% if only "DIFFERENT ADDRESS" packets are received. Probably same with "DUP". But it is only my personal point of view.

I need to think twice what to do, whether revert 5f6bec5ab57cc8beaa78f5756a0ffbdf01f28d36 or rethink counting packets (which would change exit code 2 to 0). Also the negative packet loss is strange.

May I ask you to run similar test with inetutils or busybox? I wonder if they always exit 0. 

Also, looking back at 43e38f2215a16e8a2c245e3176bea2adb3ede958, the limitation for ICMP datagram socket are:
* ident can be set on IPv6 only on SOCK_RAW.
* On IPv4 also SOCK_DGRAM can be used unless the value is 0. To set it 0, SOCK_RAW needs to be used.
=> on these limitations kernel sets the ident itself. 

> So I've played with newer ping 20221126 and ping_group_range to enforce SOCK_DGRAM usage. With SOCK_DGRAM ping utility behaves differently. In one terminal I've run ping -v 10.0.0.1 and ensured it is SOCK_DGRAM socket type. In a second terminal I've sniffed for ident (because ident is not printed in -v mode) and then run ping -v -e $ident 10.0.0.5. The ping in 1st terminal did not print DIFFERENT ADDRESS, only stopped to print anything just like with packet loss. I have no conclusion, just want to mention that.

Interesting. The second terminal started to print packets right?
@rumvadim any ideas about this issue? And generally, is there a way to avoid collisions on high traffic?

**Comment by pevik** [comment]
I also wonder if we could have option to use PID instead of our random value. It'd be less secure (it was changed to not show ping's PID), but IMHO it would help in your case because kernel would guarantee that PIDs are unique.

> ```
> `for i in $(seq 1 1000); do builddir/ping/ping ... >>/tmp/ping.log; done
> ```
BTW with this you just search for the duplicity in the logs, right? Because this does not run ping in parallel (unless you put `&` in the end). So there would be duplicity even on PID based ident. But Nagios runs ping in parallel.

**Comment by milo** [comment]
> I also wonder if we could have option to use PID instead of our random value. It'd be less secure (it was changed to not show ping's PID), but IMHO it would help in your case because kernel would guarantee that PIDs are unique.

PID is unique, but unfortunately it is not 16 bits anymore so `& 0xFFFF` would be the same, but it would probably help. Even last 16 bits of PID rotation is quite large. Actually I can already set it by `-e`.

> > ```
> > `for i in $(seq 1 1000); do builddir/ping/ping ... >>/tmp/ping.log; done
> > ```
> 
> BTW with this you just search for the duplicity in the logs, right? Because this does not run ping in parallel (unless you put `&` in the end). So there would be duplicity even on PID based ident. But Nagios runs ping in parallel.

Yes, only searched in log. I don't think there would be duplicities since every ping runs as a new process.

**Comment by metan-ucw** [comment]
I spend some time trying to figure out how to fix this properly, but it seems that the only option is to try to have a better random source, and that helps just a little bit.

One possible solution I came up with would be having a global counter that would be atomically incremented by each ping process. That would make sure that the IDs are different as long as we do not roll over too quickly. However the problem is where to put such counter. I cannot think of where to put it so that the solution would work with chroot and containers. So this won't work.

In the end we are stuck with random generator. The main problem here is that this is a variant of birthday paradox and as such collisions will certainly happen as long as we run just more than a few ping instances. I did some quick math and if I'm not mistaken the probability of at least one collision for N ping processes is:

```
N   Probability
10  0.0007
20  0.003
50  0.019
100 0.07
200 0.26
300 0.50
400 0.70
500 0.85
```

[Try other values for yourself](https://www.wolframalpha.com/input?i=1+-+65535%21%2F%28%2865535+-+500%29%21+*+65535%5E500%29)

It's clear that as long as the system runs about 500 ping processes at the same time the collision is pretty much inevitable with 85% probability.

I think that the solution here is to:
- limit Nagios so that it runs ping in small batches, e.g. 20 at a time
- check that ping tries it's best when creating random numbers
  (looking at the code the srand callback is probably not optimal)

That would limit the probability of the collision to be less than 1% however the collisions will still rarely happen, there is nothing we can do about that.

**Comment by pevik** [comment]
> * check that ping tries it's best when creating random numbers
>   (looking at the code the srand callback is probably not optimal)

Why `srand()` is problematic? We try to get good seed for it in our `iputils_srand()` wrapper (see below). It calls `getrandom()` without `GRND_RANDOM` flag, so that `/dev/urandom` is used instead of `/dev/random`. I suppose it's due noblock (we also specify `GRND_NONBLOCK` flag). BTW this code was written in ad0256f3543a69ff0d44bca838bade6aac04bcbd by Semi Kerola, util-linux use similar code: https://git.kernel.org/pub/scm/utils/util-linux/util-linux.git/tree/lib/randutils.c.

I'm not sure how much would using `/dev/random` helped (it's supposed to be kind of TRNG with limited content). If yes, there is a way how to use it in non-blocking mode (`getrandom()` + `GRND_RANDOM` + `GRND_NONBLOCK`, see `Comparison between getrandom, /dev/urandom, and /dev/random` in [`man random(7)`](https://man7.org/linux/man-pages/man7/random.7.html), we could on `EAGAIN` fallback to `/dev/urandom`).

But I would hesitate myself, if it'd be good to by default waste `/dev/random` content just for ping command. And if we hide it behind yet another getopt switch (there is really not many letters left, so it will not be any intuitive letter) and nobody will knows about it.

https://github.com/iputils/iputils/blob/0cc6da796b9a64113152c071088701cb95a72ae8/ping/ping.c#L366

https://github.com/iputils/iputils/blob/0cc6da796b9a64113152c071088701cb95a72ae8/iputils_common.c#L121-L152

**Comment by pevik** [comment]
> ... Doing (`getpid() & 0xFFFF`) ...

You probably have noticed, but in case you don't: since 5026c2221a15bf13e601eade015c971bf07a27e9 `ping` does not use `getpid()` due security reasons.

> ... I guess that we can extract some randomness from `clock_gettime()` and mix that in with xor or something along the lines to make things better. ...

Interesting, it's at least worth of trying.

**Comment by metan-ucw** [comment]
I did some more measurements and if we add some randomness to the ID generator it ends up similar to the graph above. I suppose that we cannot do actually better with a random generator here, once you start randomizing the ID generator, the distribution starts skewing so it looks like the picture above. I cannot think of other tricks, we either need to store some information about which IDs have been allocated somewhere, or we have to live with collisions.

**Comment by pevik** [comment]
Thanks a lot for the math analysis. I wonder if we should introduce a switch, which ignores `DIFFERENT ADDRESS` and `DUP` (suggested in https://github.com/iputils/iputils/issues/489#issuecomment-1728967782). It's not a solution, it's a masking of the problem, but some people may prefer it.

Also, what about the negative packet loss? (https://github.com/iputils/iputils/issues/489#issue-1902774042). Should it be kept to detect the problem?

**Comment by metan-ucw** [comment]
The more I was thinking about the problem, the more I realized that using pids was the right solution because we do not want to use random numbers that are bound to have collisions, we want ID allocation. Note that pids are unique and never reused as long as the ping program runs, this makes them ideal for such IDs. What exact problem do we have with pids being used as the ID? What attack vector are we trying to mitigate?

**Comment by pevik** [comment]
> ... using pids was the right solution ...

Maybe.

> Note that pids are unique

As Miroslav already noted (https://github.com/iputils/iputils/issues/489#issuecomment-1731141959), pids unique, but unfortunately it is not 16 bits anymore. I.e. no matter what you number you operate with `& 0xFFFF`. Or did we overlook anything?

**Comment by metan-ucw** [comment]
@pevik Well pids may be 32bits now, however the default for pid_max is still 16bit signed integer maximum, hence on most systems effectively pids are 15 bit long. And even if you raise the limit, it takes quite a lot of time to roll over to a value that would collide, so the probability is much lower than with random numbers, especially in cases where the ping commands run for a short period of time. And yes the most proper solution would be to adjust the the id to 32bit or even 64bit.

**Comment by pevik** [comment]
@milo Have you actually tried to run ping in nagios with reverted 5026c2221a15bf13e601eade015c971bf07a27e9. metan's arguments sounds reasonable.

@teknoraver You're the author of 5026c2221a15bf13e601eade015c971bf07a27e9, was there a real problem?

**Comment by pevik** [comment]
> And yes the most proper solution would be to adjust the the id to 32bit or even 64bit.

@metan-ucw That's a ICMP protocol change, right?

**Comment by metan-ucw** [comment]
@pevik Looks like it's hardwired to 16bits even in ICMPv6. I suppose that we may add an option to store 128bit random value in the optional data and if that option is on check both ident and the random cookie, however I do not think that we should inflate ping size by default.

**Comment by metan-ucw** [comment]
So all in all I would be for reverting the change that added randomization for the ident. If we want to obfuscate it a bit we can for instance add an random offset to all ping pids on the system, that should keep the properties of the sequence.

Something as:
- check if there is a /run/ping file
-  if yes, read number from the file and add it to the pid
-  if not generate random number and store it into /run/ping file and add it to the pid

There is a tricky part with creating a file with a content atomically, I guess that is done with rename(), i.e. write a temp file and attempt to rename it to a final destination with noreplace option, if that fails someone was quicker.

**Comment by teknoraver** [comment]
> To go back to raw socket you run:
> 
> ```
> sysctl net.ipv4.ping_group_range="0 1"
> ```
> 
> Obviously setup done like this is just until reboot (it needs to go to some file in `/etc/sysctl.d/*.conf`).

Nope, to disable the ICMP kernel socket you must supply an invalid range, like: `"1 0"`

**Comment by teknoraver** [comment]
> @teknoraver You're the author of [5026c22](https://github.com/iputils/iputils/commit/5026c2221a15bf13e601eade015c971bf07a27e9), was there a real problem?

The problem arised when the reply came from a different ip address than the destination. That could e caused by several problems, so it's worth knowing it.

**Comment by milo** [comment]
> @milo Have you actually tried to run ping in nagios with reverted 5026c22. metan's arguments sounds reasonable.

I did not. I'll try to elaborate why.

Nagios uses own check_ping C binary to chcek that host is alive. But this binary is only a iputils ping wrapper which parses its output. And since `DIFFERENT ADDRESS` responses are shown now and counted into statistics, the stdout parser works in a wrong way and wrongly shows that host is alive.

I've spent many hours to decode this issue and I had to move on. So I've implemented my own ping. In it I genereate random string in a packet body and on reply, I strictly compare ID, SEQ and this random string, because as RFC states, ICMP echo reply must copy data content of echo request. With this implementation I don't care about duplicit or different address reply, they are ignored and I'am 100 % sure, that probed host is alive or not.

**Comment by pevik** [comment]
> > @milo Have you actually tried to run ping in nagios with reverted [5026c22](https://github.com/iputils/iputils/commit/5026c2221a15bf13e601eade015c971bf07a27e9). metan's arguments sounds reasonable.
> 
> I did not. I'll try to elaborate why.
> 
> Nagios uses own check_ping C binary to chcek that host is alive. But this binary is only a iputils ping wrapper which parses its output. And since `DIFFERENT ADDRESS` responses are shown now and counted into statistics, the stdout parser works in a wrong way and wrongly shows that host is alive.
> 
> I've spent many hours to decode this issue and I had to move on. So I've implemented my own ping. In it I generate random string in a packet body and on reply, I strictly compare ID, SEQ and this random string, because as RFC states, ICMP echo reply must copy data content of echo request. With this implementation I don't care about duplicate or different address reply, they are ignored and I'am 100 % sure, that probed host is alive or not.

Thanks for info. Maybe reverting 5026c2221a15bf13e601eade015c971bf07a27e9 would solve your problems, but I understand you wanted to have the problem solved. Your testing of revert 5026c2221a15bf13e601eade015c971bf07a27e9 would help, but I understand if you don't have time for it.

We also were thinking about similar approach (store 128bit random value in the optional data which Cyril proposed https://github.com/iputils/iputils/issues/489#issuecomment-1778850715), but that would have to wait after the release (I'd like to release iputils in 2 or 3 weeks). Before the release I probably revert 5026c2221a15bf13e601eade015c971bf07a27e9, unless anybody raises a strong reason not to do it.

**Comment by metan-ucw** [comment]
@pevik maybe another idea for a future would be addition of a machine readable format such as JSON, so that there is no need to parse the human readable output...

**Comment by pevik** [comment]
> @pevik maybe another idea for a future would be addition of a machine readable format such as JSON, so that there is no need to parse the human readable output...

I thought there was a ticket, but we just discussed it in another ticket. Thus created: https://github.com/iputils/iputils/issues/496.

**Comment by milo** [comment]
> Thanks for info. Maybe reverting [5026c22](https://github.com/iputils/iputils/commit/5026c2221a15bf13e601eade015c971bf07a27e9) would solve your problems, but I understand you wanted to have the problem solved. Your testing of revert [5026c22](https://github.com/iputils/iputils/commit/5026c2221a15bf13e601eade015c971bf07a27e9) would help, but I understand if you don't have time for it.

Reverted, compiled, set as a Nagios check host alive command and set `net.ipv4.ping_group_range = "1 0"`. I'll let you know the bahaviour after a few days.

**Comment by milo** [comment]
With `SOCK_RAW`, so `net.ipv4.ping_group_range="1 0"`, I got 3 collisions in 3 hours. But these are between 2 tools. Nagios (ping) and Smokeping ([fping](https://github.com/schweikert/fping/blob/v5.1/src/fping.c)) running on the same server. I guess, these were always there but muted by ping before `DIFFERENT ADDRESS` commit.

```
/usr/local/bin/nagios/ping: sock4.fd: 3 (socktype: SOCK_RAW), sock6.fd: 4 (socktype: SOCK_RAW), hints.ai_family: AF_UNSPEC

ai-ai_family: AF_INET, ai-ai_canonname: 10.13.aa.bb
PING 10.13.aa.bb (10.13.aa.bb) 56(84) bytes of data.
[1698339224.290332] 64 bytes from 10.13.aa.bb: icmp_seq=1 ident=45821 ttl=64 time=0.185 ms
[1698339224.453928] 172 bytes from 172.21.bb.cc: icmp_seq=2 ident=45821 ttl=64 time=0.235 ms (DIFFERENT ADDRESS!)
[1698339225.292086] 64 bytes from 10.13.aa.bb: icmp_seq=2 ident=45821 ttl=64 time=0.184 ms
[1698339225.453665] 172 bytes from 172.21.bb.cc: icmp_seq=3 ident=45821 ttl=64 time=0.231 ms (DIFFERENT ADDRESS!)
[1698339226.293786] 64 bytes from 10.13.aa.bb: icmp_seq=3 ident=45821 ttl=64 time=0.152 ms

--- 10.13.aa.bb ping statistics ---
3 packets transmitted, 5 received, -66.6667% packet loss, time 2003ms
rtt min/avg/max/mdev = 0.152/0.197/0.235/0.031 ms
```

After that I switched to `SOCK_DGRAM`, so `net.ipv4.ping_group_range="0 4294967294"` and I didn't received any `DIFFERENT ADDRESS` response in 14 hours. I guess, kernel handles it differently.

**Comment by pevik** [comment]
Sou

> @pevik maybe the kernel already checks it?
> 
> Maybe here: https://elixir.bootlin.com/linux/v6.6.4/source/net/ipv4/ping.c#L110

Yes, it looks like kernel does. 

Change is from https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=6d0bfe22611602f36617bc7aa2ffa1bbb2f54c67 (from v3.11-rc1). And I guess `ping.c` is `[net/ipv4/ping.c](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/net/ipv4/ping.c?id=6d0bfe22611602f36617bc7aa2ffa1bbb2f54c67)`, not `ping.c` from our userspace ping.
